### PR TITLE
.Net: Handlebars syntax example 

### PR DESCRIPTION
**Pull request by joslat**
This example shows how to use Handlebars different syntax options

### Motivation and Context
This example shows how to use Handlebars different syntax options like:
  - set (works)
  - function calling (works)
  - loops (each) - not working at the moment, but tries to do so over a complex object generated by one of the functions: a JSON array-
  - array - not working at the moment - to accumulate the results of the loop in an array
  - conditionals (works)
  - concatenation (works)

In order to create a Prompt Function that fully benefits from the Handlebars syntax power.
 The example also shows how to use the HandlebarsPlanner to generate a plan (and persist it) which was used to generate the initial Handlebar template.
The example also shows how to create two prompt functions and a plugin to group them together.

  1. Why is this change required?
     -  there is no example that shows anything but the simple syntax of handlebars
  2. What problem does it solve?
     - help others understand handlebars and how to do "all the things" possible ;)
  4. What scenario does it contribute to?
     - Learning Semantic Kernel and enabling its usage.
  5. If it fixes an open issue, please link to the issue here.
     - issue: no handlebars example with sets, function calling, conditionals and loops. ;)

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by joslat** [comment]
@matthewbolanos - if you want to go over this I am in anytime.

File: dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+//  - conditionals (works)
+//  - concatenation (works)
+
+// In order to create a Prompt Function that fully benefits from the Handlebars syntax power.
+// The example also shows how to use the HandlebarsPlanner to generate a plan (and persist it) which was used to generate the initial Handlebar template.
+// The example also shows how to create two prompt functions and a plugin to group them together.
+public static class Example77_HandlebarsPromptSyntax
+{
+    public static async Task RunAsync()
+    {
```
Comment (Krzysztof318) [inline dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:24]: Examples has been changed to integration tests today, look how it is done currently in repo.

**Review by Krzysztof318 — CHANGES_REQUESTED** [review]

File: dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs (hunk 1 above)
Comment (joslat) [reply dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:29]: Thanks, @Krzysztof318! that's done - now it just needs a check/review in regards of the Handlebars template as discussed with @matthewbolanos in Discord - https://discord.com/channels/1063152441819942922/1194728332139442337/1196839845851582594  :)

Comment (Krzysztof318) [reply dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:29]: @joslat Hi, one more small fix yet :) Change Console to this.WriteLine. WriteLine from BaseTest is using ITestOutputHelper interface from xunit

**Comment by joslat** [comment]
"Maybe" the issue I am having with the loops is due to this issue: https://github.com/microsoft/semantic-kernel/issues/4596

**Comment by joslat** [comment]
I did not realize I had to "resolve" the conversation - just did...

**Review by markwallace-microsoft — DISMISSED** [review]

**Review by markwallace-microsoft — COMMENTED** [review]
Can you fix the typo please

**Comment by markwallace-microsoft** [comment]
@joslat Thanks for the contribution, this is a useful addition. Can you please fix the typo and formatting changes.

**Comment by joslat** [comment]
> @joslat Thanks for the contribution, this is a useful addition. Can you please fix the typo and formatting changes.

Thanks  @markwallace-microsoft  - Unsure you want to accept the PR as it is as mentions a couple of things that did not work - haven't seen if they are fixed though. 

I put some code in there that is meant to be left as an example on how to do things/use handlebars... you tell me if you want me to adapt, remove any section... we can have a sync or you can tell me here - or go and do it.

File: dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            {{/if}}";
+
+        /// did not work:
+        /// {{set ""outputDescription"" (concat ""PRODUCT "" (Add @index 1) "": "" compellingDescription)}}
+        /// Reason: helper add not found
+        /// Also the array thing doesn't seem to work
+        ///     {{set ""finalDescriptions"" (array)}}
+        /// what fails on the each loop, when I use it I set generateEngagingDescriptions and make the prompt function output
+        /// JSON to iterate through it. Technically it is the proper syntax but somehow it fails. some returns from the LLM are
+        /// absolute hallucinations :/P
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:216]: Can you create an issue for the failures

File: dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        var HandlebarsSPromptFunction = kernel.CreateFunctionFromPrompt(
```
Comment (markwallace-microsoft) [inline dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:217]: Should this be `handlebarsPromptFunction`?

**Comment by joslat** [comment]
@markwallace-microsoft  I also separated and put the handlebars syntax in a more convenient way, showing three progressively more complex handlebars syntax examples.
The one worrying me is the last issue, same prompt function generates different output inside handlebars prompt. I even put an example syntax and is flawless when working on its own. Really puzzled here.
But it is working and a good example, runs but provides some hallucination as final response - I believe there a couple to three things to get better in the planner and some strange issue (maybe it is me?) on the function call inside the Prompt template (handlebars edition) :)

**Review by markwallace-microsoft — COMMENTED** [review]
Needs passing PR checks

**Comment by markwallace-microsoft** [comment]
> @markwallace-microsoft I also separated and put the handlebars syntax in a more convenient way, showing three progressively more complex handlebars syntax examples. The one worrying me is the last issue, same prompt function generates different output inside handlebars prompt. I even put an example syntax and is flawless when working on its own. Really puzzled here. But it is working and a good example, runs but provides some hallucination as final response - I believe there a couple to three things to get better in the planner and some strange issue (maybe it is me?) on the function call inside the Prompt template (handlebars edition) :)

@joslat  Thanks for the update. I've asked @teresaqhoang to also take a look as she is working on improvements to the Handlebars planner. But this is a great start and will help developers to better understand the syntax.

**Review by teresaqhoang — CHANGES_REQUESTED** [review]
Hi @joslat, pologies for the delayed review, have been heads down on a PR and this slipped past my radar

Left some comments — in general, the function descriptions and user prompt need polishing, then the HB planner will produce more reliable plans. Model is usually pretty good if you pass in descriptions and ask it to help you fine tune them!

**Comment by joslat** [comment]
Hi @teresaqhoang, have resolved some of your comments but unsure the implementation is ideal - so I have not labelled them as "resolved" - also would love to get samples on two of your suggestions, if you have any code links to what you suggest it would be perfect.
I still have pending the last suggestion but I'll do that a bit after. 
Thanks for the great feedback - I give you the🥇 for that!!

Also, those issues might be affecting/impacting the outcome/function of this sample:

- https://github.com/microsoft/semantic-kernel/issues/4895
- https://github.com/microsoft/semantic-kernel/issues/4892 @markwallace-microsoft  I understand that this issue by you means that a JSON parsing will not work once this is fixed, right?
- https://github.com/microsoft/semantic-kernel/issues/4895 - is this something complex/hard to solve?
- https://github.com/microsoft/semantic-kernel/issues/4870 @teresaqhoang  according to this it seems OpenAI is best for "planning things" - thoughts? - i am usually using Microsoft OpenAI services as I've got an Azure Sponsorship subscription... I also have an OpenAI paid subscription and paid credits too so can do both - is one better than the other at the moment?
- https://github.com/microsoft/semantic-kernel/issues/4721 - I guess this is triggering your handlebars enhancements @markwallace-microsoft - yes? or is this another separate issue?

**Comment by joslat** [comment]
Any news on this? :)

**Comment by teresaqhoang** [comment]
> Hi @teresaqhoang, have resolved I believe all of your comments but unsure the implementation is ideal (or the prompt improvement suggestions are fully applied) - so I have not labelled them as "resolved" - also would love to get samples on two of your suggestions, if you have any code links to what you suggest it would be perfect. Thanks for the great feedback - I give you the🥇 for that!!
> 
> Also, those issues might be affecting/impacting the outcome/function of this sample:
> 
> * [.Net: Same plugin behaves differently inside a prompt template execution. #4895](https://github.com/microsoft/semantic-kernel/issues/4895)
> * [.Net: Enhancements to Handlebars syntax #4892](https://github.com/microsoft/semantic-kernel/issues/4892) @markwallace-microsoft  I understand that this issue by you means that a JSON parsing will not work once this is fixed, right?
> * [.Net: Same plugin behaves differently inside a prompt template execution. #4895](https://github.com/microsoft/semantic-kernel/issues/4895) - is this something complex/hard to solve?
> * [.Net Planners with models other than openai #4870](https://github.com/microsoft/semantic-kernel/issues/4870) @teresaqhoang  according to this it seems OpenAI is best for "planning things" - thoughts? - i am usually using Microsoft OpenAI services as I've got an Azure Sponsorship subscription... I also have an OpenAI paid subscription and paid credits too so can do both - is one better than the other at the moment?
> * [.Net: Handlebars Planner Failing to Parse JSON Array #4721](https://github.com/microsoft/semantic-kernel/issues/4721) - I guess this is triggering your handlebars enhancements @markwallace-microsoft - yes? or is this another separate issue?

Hey @joslat, 

Apologies for the delayed response - I haven't been getting notifications for changes on this PR for some reason, will be more diligent in checking the PR directly moving forward.

Thanks for iterating! I can happily help you with more guidance! Can you reply on the unresolved conversation threads directly with your questions / points of concerns? And also call out the ones you want code samples for? 

I'll also pull this example and test it locally to see if I can better identify some pain points the model's hitting.

**Comment by joslat** [comment]
@teresaqhoang  "Pong" ;) - ball is in your ground :D 
Let me know if there is anything more I can help here, happy to...

File: dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+                {{!-- Step 4: Print all product names and compelling descriptions --}}
+                OUTPUT The following product descriptions as is, do not modify anything:
+                {{json finalDescriptionsV2}}
+    
+            {{else}} 
+                {{!-- Example of concatenating text and variables to finally output it with json --}}
+                {{set ""finalOutput"" (concat ""Description 1: "" productNames "" Description 2: "" productNames2)}}
+                {{json finalOutput}}
+            {{/if}}";
```
Comment (teresaqhoang) [inline dotnet/samples/KernelSyntaxExamples/Example77_HandlebarsPromptSyntax.cs:114]: In terms of showcase a HB prompt template, these are much too complex. 

Example of how you can reduce complexity:
```handlebars
{{!-- Omit the initial variable mapping of inputs (this is required for the planner to reduce error) but as a dev executing their own HB template, you should have full context over how input are maps. --}}
{{set ""productNames"" (productMagician-GenerateJSONProducts input)}}

{{#if generateEngagingDescriptions}} 
{{!-- Since SK doesn't support mutable arrays, you must concatenate each result onto a string. --}}
    {{set ""finalProducts"" """"}}

    {{#each productNames}}
{{!-- Refactor your GenerateJSONProducts function to return just an array of products, then remove the inner {{#each}}} block. It's unnecessarily complex. --}}
        {{set ""productDescription"" (concat ""Product Name: "" this.name "" Description: "" this.description)}}
        {{set ""compellingDescription"" (productMagician-GenerateProductCompellingDescription this)}}
        {{set ""outputProduct"" (concat ""PRODUCT :"" this.name "" Engaging Description: "" compellingDescription)}}

{{!-- Check if you're on the first object; if not, concatenate to end of string.  --}}
        {{#if @first}}{{set ""finalProducts"" outputProduct}}
        {{else}}{{set ""finalProducts"" (concat finalProducts ""\n\n"" outputProduct)}}
        {{/if}}
    {{/each}}
{{!--  OUTPUT The following product descriptions as is, do not modify anything --}}   
    {{json finalProducts}}
{{else}} 
{{!-- Return products as is since the conditional is checking whether descriptions should be regenerated. --}}
    {{json productNames}}
{{/if}}
```

**Review by teresaqhoang — CHANGES_REQUESTED** [review]

**Comment by teresaqhoang** [comment]
@joslat I see you resolved some of the conversations (thanks for that) but I'm not seeing any replies to the conversations that are still open so I'm not sure what kind of clarity you need where. I tried to reply best I could.

Do you see the option to reply directly on those conversation threads? If not, we can try to troubleshoot your access issues
image

IMO I think this example is too complex for what it's trying to showcase.

We try to keep the Syntax examples as simple as possible to help users grasp the concept easier, and I think this one (since it's derived from the planner) can be hard to follow. 

I left more comments with suggestions on how to simplify these samples

**Comment by teresaqhoang** [comment]
> @joslat I see you resolved some of the conversations (thanks for that) but I'm not seeing any replies to the conversations that are still open so I'm not sure what kind of clarity you need where. I tried to reply best I could.
> 
> Do you see the option to reply directly on those conversation threads? If not, we can try to troubleshoot your access issues image
> 
> IMO I think this example is too complex for what it's trying to showcase.
> 
> We try to keep the Syntax examples as simple as possible to help users grasp the concept easier, and I think this one (since it's derived from the planner) can be hard to follow.
> 
> I left more comments with suggestions on how to simplify these samples
> 
> Also happy to schedule a meeting so we can parallel program this together if you'd like :)

> users

Hey @joslat, I discussed your PR with the team today and we think there's a lot of value in examples shown here, but some ideas are getting lost in translation and complexity. 

What do you think about pivoting this example to show the difference between Handlebars Plan and a Handlebars Template? It would still fit your goal of showing the Handlebars syntax.

Some differences:
- a Handlebars Plan runs a sequence of steps to achieve a defined goal, so it can be more dynamic and doesn't produce concrete results. When a HB plan is invoked, the template is simply rendered using a HB instance, and the raw render result is the output.
- whereas a pure Handlebars Prompt template is invoked as a Semantic function using SK - a template can call multiple functions within it, but when a HB prompt template is invoked, it's first rendered using a HB instance and then SK passes the render result to the LLM. Then the completion result is the final output.

I created an example here that refactors your `GenerateProductCompellingDescription` function into a new semantic function called `GenerateDetailedProducts` that leverages a HB prompt template:
https://github.com/teresaqhoang/semantic-kernel/commit/78bb77bd3c54d993aefb69295f635847ec6e88a7

I left some prompt engineering pointers directly on that commit. Let me know if you have any questions

**Comment by joslat** [comment]
Hi @teresaqhoang would be my honor to do this in "extreme programming" way, that would be fantastic for "getting it done" - also this would ensure the focus in showcasing as well the differences between Handlebars Plan and  Handlebars Template.

**Comment by teresaqhoang** [comment]
Hi @joslat, @matthewbolanos will handle follow-up on this moving forward! Let me know if I can help provide code reviews.

**Comment by markwallace-microsoft** [comment]
@joslat Thanks for the PR. We have decided we're not going to invest in Handlebars for planning and instead focus on providing first class support for planning with Function Calling and Python code execution using the new Azure Container Applications functionality for serverless code interpreter sessions.

**Comment by joslat** [comment]
That makes a lot of sense, but will we be able - read supported - to run python on other places like docker, locally if we dare as well?
Will this mean that the StepWisePlanner V2 will also be deprecated? I know it is mostly a wrapper over Function Calling and theoretically better...
